### PR TITLE
fix #6833 improve feature mismatch metric

### LINKED ISSUE #6833: Regression in conda 4.3.33 regarding build features
It seems that `conda 4.3.33` installs packages that have a build feature when installing any package that has the same build feature.

From a fresh miniconda, with conda pinned at `4.3.32`:

```sh
scorlay@mbp ~]$ conda install cryptopp -c QuantStack -c conda-forge
Fetching package metadata ...............
Solving package specifications: .

Package plan for installation in environment /Users/scorlay/miniconda3:

The following NEW packages will be INSTALLED:

    cling-patches: 1-0           QuantStack
    cryptopp:      5.6.5-cling_0 QuantStack [cling]
```

With `conda 4.3.33`

```sh
[scorlay@mbp ~]$ conda install cryptopp -c QuantStack -c conda-forge
Fetching package metadata ...............
Solving package specifications: .

Package plan for installation in environment /Users/scorlay/miniconda3:

The following NEW packages will be INSTALLED:

    clangdev:      5.0.0-cling_1 QuantStack  [cling]
    cling-patches: 1-0           QuantStack 
    cryptopp:      5.6.5-cling_0 QuantStack  [cling]
    icu:           58.2-0        conda-forge
    libiconv:      1.15-0        conda-forge
    libxml2:       2.9.7-0       conda-forge
    llvmdev:       5.0.0-cling_1 QuantStack  [cling]
```

The only package tracking the `cling` build feature is `cling-patches`.

Notice in the second case that `clangdev` gets installed while it is not a dependency.

### PR DESCRIPTION
**Pull request by mcg1969**
Attempts to fix #6833. Will copy the relevant text from there

### CODE REVIEW

**Comment by mcg1969** [comment]
That said: I have an idea for another approach to the `generate_feature_metric` method; @mbargull you may want to try this. The idea is this. Right now the two feature optimization passes are:

1. Minimize the number of active features activated.
2. Maximize the number of packages installed that contain an active feature.

The problem has been with step 2 there, and how it has been implemented. Indeed, as I've stated it here, it's quite obvious why it's causing the problems @SylvainCorlay is seeing now. But I think a better metric is this:

2. Minimize the number of packages installed with _feature mismatch_. That is: the package _has_ a featured alternative, but it was not installed for other reasons (say, dependency conflicts).

Here is the code I've written to generate this optimization metric. @mbargull it doesn't break the test I created for PR #6766. And I _believe_ this will resolve the behavior we're seeing here, because it's not doing anything to _favor_ the installation of a package you didn't ask for.

**Comment by mcg1969** [comment]
If you're reading this and wondering why conda should be _minimizing_ the number of installed features---note that if you _install_ a package that contains a `track_feature`, then it is "stuck" and won't be minimized away. So really, this is just the way for `conda` to try to ensure you don't get a feature you didn't ask for.

**Comment by mcg1969** [comment]
@mbargull can you try this out on the scenarios that you have been most keenly interested?

**Comment by mbargull** [comment]
> Right now the two feature optimization passes are: \[...\]
>
> 2. Maximize the number of packages installed that contain an active feature.
>
>\[...\] But I think a better metric is this:
>
> 2. Minimize the number of packages installed with feature mismatch. That is: the package has a featured alternative, but it was not installed for other reasons (say, dependency conflicts).

Yes, that makes much more sense, and sounds more in line with the intended use of features (as I came to understand them, at least ;).
I'll try it out in a couple of hours and will also take a look at the code. Thank you for providing an update on this so quickly!

File: conda/resolve.py [hunk 1]
```diff
@@ -624,20 +624,22 @@ def generate_update_count(self, C, specs):
 
     def generate_feature_metric(self, C):
         eq = {}  # a C.minimize() objective: Dict[varname, coeff]
-        total = 0
+        # Given a pair (dist, feature), assign a "1" score IF:
+        # - The dist is installed
+        # - The dist does NOT require the feature
+        # - At least one package in the group DOES require the feature
+        # - A package that tracks the feature is installed
```
Comment (mbargull) [inline conda/resolve.py]: Thanks for including this comment! That's very helpful to countercheck the code against.

**Comment by mcg1969** [comment]
Thanks so much for the review @mbargull. I am relieved to hear it seems to cover both regimes of difficulty.

**Comment by mbargull** [comment]
Me too. Thanks for this.

**Comment by kalefranz** [comment]
We should ensure this PR also resolves https://github.com/ContinuumIO/anaconda-issues/issues/8471

**Comment by mbargull** [comment]
It does: https://github.com/ContinuumIO/anaconda-issues/issues/8471#issuecomment-364137796

**Comment by kalefranz** [comment]
I've updated this PR with @mbargull's requested code comments, as well as expanding on some variable names to hopefully make the code more self-documenting.

@mcg1969 Are you fine removing the WIP tag and having this merged at this point?

**Comment by mcg1969** [comment]
Well, I do feel like we could benefit from some key tests. Like for instance #8471 ...

**Comment by kalefranz** [comment]
Don't disagree with that.  Who's going to write them?

**Comment by mbargull** [comment]
```python
# in test_resolve.py

def test_surplus_features():
    index = {
        'feature-1.0-0.tar.bz2': IndexRecord(**{
            'name': 'feature',
            'version': '1.0',
            'build': '0',
            'build_number': 0,
            'track_features': 'feature',
        }),
        'package1-1.0-0.tar.bz2': IndexRecord(**{
            'name': 'package1',
            'version': '1.0',
            'build': '0',
            'build_number': 0,
            'features': 'feature',
        }),
        'package2-1.0-0.tar.bz2': IndexRecord(**{
            'name': 'package2',
            'version': '1.0',
            'build': '0',
            'build_number': 0,
            'depends': ['package1'],
            'features': 'feature',
        }),
        'package2-2.0-0.tar.bz2': IndexRecord(**{
            'name': 'package2',
            'version': '2.0',
            'build': '0',
            'build_number': 0,
            'features': 'feature',
        }),
    }
    r = Resolve({Dist(key): value for key, value in iteritems(index)})
    install = r.install(['package2', 'feature'])
    assert 'package1' not in set(d.name for d in install)
```

**Comment by mbargull** [comment]
Oh, just noticed now, that I actually wanted
```diff
-        'package2-2.0-0.tar.bz2': IndexRecord(**{
+        'package2-1.0-1.tar.bz2': IndexRecord(**{
             'name': 'package2',
-            'version': '2.0',
-            'build': '0',
-            'build_number': 0,
+            'version': '1.0',
+            'build': '1',
+            'build_number': 1,
             'features': 'feature',
         }),
```
in that test case, i.e., make the decision dependent on build number and dependency only.
I'm a little surprised that even with the version number increase that test failed without the changes from this PR... @mcg1969, should we use the test as committed or rather like in the diff above? Either one fails without this PR and passes with it.

**Comment by kalefranz** [comment]
No reason not to include both.

**Comment by kalefranz** [comment]
@mcg1969 Are you good to go on this?

**Comment by kalefranz** [comment]
@mcg1969 We're bumping up against the last few hours of what we can sneak into conda for the next Anaconda and matching Miniconda.  If we want this in, we have to be moving on it.

**Comment by mbargull** [comment]
A `Miniconda3 4.3.34` with this would be nice, indeed (esp. since Bioconda/conda-forge still use `4.3.x` and I assume this could likely be the last `Miniconda 4.3.x`, right?).

**Comment by mcg1969** [comment]
Understood. Thanks for the tremendous help on this, both of you.

**Comment by mingwandroid** [comment]
Once AD 5.1 comes out with conda 4.4 it would be weird for us to make a Miniconda with 4.3.

I guess we *could* make an exception here for you guys but what's holding up conda 4.4 adoption?

Pinging @csoja and @nehaljwani.

**Comment by mbargull** [comment]
> Once AD 5.1 comes out with conda 4.4

Ah, okay, wasn't aware/sure of that.

> but what's holding up conda 4.4 adoption?

Good question. Can anyone from @conda-forge/core (`@conda-forge/core`) or @bioconda/core (`@bioconda/core`) comment on why we shouldn't update to 4.4?
IMHO, 4.4's teething troubles have been worked out in the patch releases, so from me there won't be any objections.

BTW: Can anyone confirm whether team mentions work outside of an organization? At least they are never *rendered* on GitHub... Either way, I'll put the word out on Gitter just to be sure.

**Comment by bgruening** [comment]
Nothing against 4.4 from my side. Thanks for all the work you all have put into this.

**Comment by kalefranz** [comment]
I’ve got a PR against conda-build-all that’s still failing tests for some reason. Thought conda 4.4.9 fixed it, but apparently not. That’s the biggest road block right now I think.

**Comment by kalefranz** [comment]
@mbargull if you have the bandwidth, take a look at https://github.com/conda-tools/conda-build-all/pull/96 and see if you can spot what’s going on in those failing tests. 

I thought I had fixed the issue with https://github.com/conda/conda/pull/6808, but guess I was wrong.

**Comment by mbargull** [comment]
Just opened the Travis log for that a couple of seconds ago 😉

**Comment by isuruf** [comment]
No issues with conda=4.4 for conda-forge. We'll just downgrade to conda=4.3, until conda-forge is ready.

**Pull request merged**